src/transformers/models/jamba/modeling_jamba.py

```diff
@@ -0,0 +1,2127 @@
+
+
+class JambaMambaDecoderLayer(nn.Module):
+    def __init__(self, config: JambaConfig, num_experts: int, layer_idx: int):
+        super().__init__()
+
+        self.mamba = JambaMambaMixer(config=config, layer_idx=layer_idx)
+
+        num_experts_per_tok = config.num_experts_per_tok if num_experts > 1 else 1
+        self.moe = JambaSparseMoeBlock(config, num_experts=num_experts, num_experts_per_tok=num_experts_per_tok)
+        self.input_layernorm = JambaRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+        self.pre_moe_layernorm = JambaRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value: Optional[HybridMambaAttentionDynamicCache] = None,
+        output_attentions: Optional[bool] = False,
+        output_router_logits: Optional[bool] = False,
+        use_cache: Optional[bool] = False,
+        **kwargs,
+    ) -> Tuple[torch.FloatTensor, Optional[Tuple[torch.FloatTensor, torch.FloatTensor]]]:
+        if "padding_mask" in kwargs:
+            warnings.warn(
+                "Passing `padding_mask` is deprecated and will be removed in v4.37. Please make sure use `attention_mask` instead.`"
+            )
+        """
+        Args:
+            hidden_states (`torch.FloatTensor`): input to the layer of shape `(batch, seq_len, embed_dim)`
+            attention_mask (`torch.FloatTensor`, *optional*): attention mask of size
+                `(batch, sequence_length)` where padding elements are indicated by 0.
+            past_key_value (`Tuple(torch.FloatTensor)`, *optional*): cached past key and value projection states
+            output_attentions (`bool`, *optional*):
+                Whether or not to return the attentions tensors of all attention layers. See `attentions` under
+                returned tensors for more detail.
+            output_router_logits (`bool`, *optional*):
+                Whether or not to return the logits of all the routers. They are useful for computing the router loss, and
+                should not be returned during inference.
+            use_cache (`bool`, *optional*):
+                If set to `True`, `past_key_values` key value states are returned and can be used to speed up decoding
+                (see `past_key_values`).
+        """
+
+        residual = hidden_states
+
+        hidden_states = self.input_layernorm(hidden_states)
+
+        hidden_states, present_key_value = self.mamba(
+            hidden_states=hidden_states,
+            past_key_value=past_key_value,
+        )
+        bs, seqlen, _ = hidden_states.shape
+        past_seqlen = self._get_past_seqlen(past_key_value, seqlen)
+        num_attention_heads = self.mamba.config.num_attention_heads
+        self_attn_weights = torch.empty(bs, num_attention_heads, seqlen, past_seqlen, device="meta")
```

Comment:
IMO makes more sense to just return None, I guess that is breaking something / is less easy to manipulate?

Reply:
also would allow us to remove the get past seq length so in favor of that!

Reply:
Exactly.. 
In the original implementation we returned `None`, but then many tests failed. I had the option to override a bunch of unittests or return this empty tensor on the meta device (so it wouldn't consume extra memory) just to align with the conventions in the code. 
Another option is to not return attentions also for the attention layers, and mark the model as one that doesn't return attentions (like MambaModel). But since Jamba does have attention layers, I thought that not returning attention weights for these layers is giving up on some functionality and visibility into the model.

I don't have  a hard opinion here.. we can go both ways. Which do you think is better?

Reply:
Yeah the tests can be annoying. I think the middle ground is not to return `None` and if `None` don't concatenate to the `all_attention_outputs`. If that fails, let's just skip the tests, and leave it as a feature request as we want to have this PR merged ASAP!

Reply:
OK that worked 😃 
I actually had to override less unittests after returning `None` as the attention weights from Mamba layers, but concatenating to `all_attention_outputs` only if it's not `None` (so only for Attention layers).
Now, the length of `outputs.attentions` is equal to the number of attention layers, so is smaller than `num_hidden_layers`, but all tests pass so everything should work.